server/mqtt.go

```diff
@@ -329,13 +334,26 @@ type mqttRetainedMsgRef struct {
 	sub   *subscription
 }
 
+// mqttSub contains fields associated with a MQTT subscription, and is added to
+// the main subscription struct for MQTT message delivery subscriptions. The
+// delivery callbacks may get invoked before sub.mqtt is set up, so they should
+// acquire either sess.mu or sess.subsMu before accessing it.
 type mqttSub struct {
-	qos byte
-	// Pending serialization of retained messages to be sent when subscription is registered
-	prm *mqttWriter
-	// This is the JS durable name this subscription is attached to.
+	// The sub's QOS and the JS durable name. They can change when
+	// re-subscribing, and are used in the delivery callbacks. They can be
+	// quickly accessed using sess.subsMu.RLock, or under the main session lock.
+	qos   byte
 	jsDur string
-	// If this subscription needs to be checked for being reserved. E.g. # or * or */
+
+	// Pending serialization of retained messages to be sent when subscription
+	// is registered. The sub's delivery callbacks must wait until `prm` is
+	// ready (can block on sess.mu for that, too).
+	prm      [][]byte
+	prmReady *sync.WaitGroup
```

Comment:
NS: check that gets (re-)set properly when re-subscribing and re-delivering retained (with the new QOS).

Reply:
NS: do we, should we re-deliver retained if there's a re-subscribe with the same QOS as before?

Reply:
I looked into pre-serialization, and I think I ran into a dependency I couldn't resolve easily. I'll look into it one more time.

And yes, on while this change would unblock other subs on the session, it would still be holding the ROUTE, etc. messages coming into the sub.

Reply:
Hm, ok, I am going to defend my approach some more.

1. The "serialization" no longer involves copying the message bytes, it now takes 
  - (a) a second `as.sl` scan for each subject (can be cached from loadRetainedMessages if we wanted to)
  - (b) serializing the header (QOS, PI) that is specific to the sub/message
 Both of these should already be quite quick.

2. Using a separate (from `sess.mu`), cheaper lock for initializing `sub.mqtt` just so we can even engage in the (QOS0) callback - makes sense, thus `subsMu`. 

3. 0/5 about `sub.mqtt.prmReady` - it can be removed, because (a). We can use sess.subsMu (just like sess.mu before) to block for the duration of building the `prm`.